File: Java/ChainOfResponsibility/CashDispenser/src/com/ericgtkb/CashDispenser.java
package com.ericgtkb;

public class CashDispenser {
    private DispenseChain chain;

    public CashDispenser() {
        chain = new DispenseChain100Dollar();
        DispenseChain chain50 = new DispenseChain50Dollar();
        DispenseChain chain20 = new DispenseChain20Dollar();
        DispenseChain chain5 = new DispenseChain5Dollar();
        DispenseChain chain1 = new DispenseChain1Dollar();

        chain.setNextChain(chain50);
        chain50.setNextChain(chain20);
        chain20.setNextChain(chain5);
        chain5.setNextChain(chain1);
    }

    public void dispense(int amount) {
        System.out.println("==================================================");
        System.out.println("Dispensing cash of the amount $" + amount + "...");
        chain.dispense(amount);
        System.out.println();
    }
}

File: C++/ChainOfResponsibility/CashDispenser/cash_dispenser.cc
#include <iostream>
#include <string>
#include "cash_dispenser.h"


CashDispenser::CashDispenser() : chain(), chain50(), chain20(), chain5(), chain1() {
    chain.set_next_chain(&chain50);
    chain50.set_next_chain(&chain20);
    chain20.set_next_chain(&chain5);
    chain5.set_next_chain(&chain1);
}

void CashDispenser::dispense(int amount) {
    std::cout << std::string(50, '=') << '\n';
    std::cout << "Dispensing cash of the amount $" << amount << "...\n";
    chain.dispense(amount);
    std::cout << '\n';
}
